File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/patient/actions/SEBronchoconstriction.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

package mil.tatrc.physiology.datamodel.patient.actions;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.BronchoconstrictionData;
import mil.tatrc.physiology.datamodel.bind.EnumOnOff;
import mil.tatrc.physiology.datamodel.properties.SEScalar0To1;

public class SEBronchoconstriction extends SEPatientAction
{
	protected SEScalar0To1 severity;
	
	public SEBronchoconstriction()
	{
		severity = null;
	}
	
	public void copy(SEBronchoconstriction other)
	{
		if(this==other)
			return;
		super.copy(other);
		if (other.severity != null)
			getSeverity().set(other.getSeverity());
		else if (severity != null)
			severity.invalidate();
	}
	
	public void reset()
	{
		super.reset();
		if (severity != null)
			severity.invalidate();
	}
	
	public boolean isValid()
	{
		return hasSeverity();
	}
	
	public boolean load(BronchoconstrictionData in)
	{
		super.load(in);
		getSeverity().load(in.getSeverity());
		return isValid();
	}
	
	public BronchoconstrictionData unload()
	{
		BronchoconstrictionData data = CDMSerializer.objFactory.createBronchoconstrictionData();
		unload(data);
		return data;
	}
	
	protected void unload(BronchoconstrictionData data)
	{
		super.unload(data);
		if (severity != null)
			data.setSeverity(severity.unload());
	}
	
	public boolean hasSeverity()
	{
		return severity == null ? false : severity.isValid();
	}
	public SEScalar0To1 getSeverity()
	{
		if (severity == null)
			severity = new SEScalar0To1();
		return severity;
	}
	
	public String toString()
	{
		if (severity != null)
			return "Bronchoconstriction" 
					+ "\n\tSeverity: " + getSeverity();
		else
			return "Action not specified properly";
	}
}

File: projects/biogears/libBiogears/src/cdm/patient/actions/SEBronchoconstriction.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

#include <biogears/cdm/patient/actions/SEBronchoconstriction.h>
#include <biogears/cdm/properties/SEScalar0To1.h>

namespace biogears {
SEBronchoconstriction::SEBronchoconstriction()
  : SEPatientAction()
{
  m_Severity = nullptr;
}

SEBronchoconstriction::~SEBronchoconstriction()
{
  Clear();
}

void SEBronchoconstriction::Clear()
{
  SEPatientAction::Clear();
  SAFE_DELETE(m_Severity);
}

bool SEBronchoconstriction::IsValid() const
{
  return SEPatientAction::IsValid() && HasSeverity();
}

bool SEBronchoconstriction::IsActive() const
{
  return IsValid() ? !m_Severity->IsZero() : false;
}

bool SEBronchoconstriction::Load(const CDM::BronchoconstrictionData& in)
{
  SEPatientAction::Load(in);
  GetSeverity().Load(in.Severity());
  return true;
}

CDM::BronchoconstrictionData* SEBronchoconstriction::Unload() const
{
  CDM::BronchoconstrictionData* data(new CDM::BronchoconstrictionData());
  Unload(*data);
  return data;
}

void SEBronchoconstriction::Unload(CDM::BronchoconstrictionData& data) const
{
  SEPatientAction::Unload(data);
  if (m_Severity != nullptr)
    data.Severity(std::unique_ptr<CDM::Scalar0To1Data>(m_Severity->Unload()));
}

bool SEBronchoconstriction::HasSeverity() const
{
  return m_Severity == nullptr ? false : m_Severity->IsValid();
}
SEScalar0To1& SEBronchoconstriction::GetSeverity()
{
  if (m_Severity == nullptr)
    m_Severity = new SEScalar0To1();
  return *m_Severity;
}

void SEBronchoconstriction::ToString(std::ostream& str) const
{
  str << "Patient Action : Bronchoconstriction";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  str << "\n\tSeverity: ";
  HasSeverity() ? str << *m_Severity : str << "NaN";
  str << std::flush;
}
}